File: JS/core/arrayex.js
import { iterate, rangeNumber, map, delay } from "../.fable/fable-library.3.0.0/Seq.js";
import { equals } from "../.fable/fable-library.3.0.0/Util.js";
import { permute } from "../.fable/fable-library.3.0.0/Array.js";

function swapIndexPairs(fRand, lst) {
    return delay(() => map((i) => [i, fRand(i + 1)], rangeNumber(lst.length - 1, -1, 1)));
}

function swapIndexMap(index, fromIndex, toIndex) {
    if (equals(index, fromIndex)) {
        return toIndex;
    }
    else if (equals(index, toIndex)) {
        return fromIndex;
    }
    else {
        return index;
    }
}

function swap(indexPair_0, indexPair_1, lst) {
    const indexPair = [indexPair_0, indexPair_1];
    return permute((index) => swapIndexMap(index, indexPair[0], indexPair[1]), lst);
}

export function shuffleInPlace(fRand, lst) {
    const swap_1 = (pair) => {
        const temp = lst[pair[0]];
        lst[pair[0]] = lst[pair[1]];
        lst[pair[1]] = temp;
    };
    iterate(swap_1, swapIndexPairs(fRand, lst));
    return lst;
}

File: src/core/arrayex.fs
module Array

let inline of1 a = [| a |]
let inline of2 a b = [| a; b |]
let inline of3 a b c = [| a; b; c |]
let inline of4 a b c d = [| a; b; c; d |]
let inline of5 a b c d e = [| a; b; c; d; e |]

let private swapIndexPairs fRand lst =
  seq
    {
      for i in (Array.length lst - 1) .. -1 .. 1 do
        yield (i, fRand (i + 1))
    }

let private swapIndexMap index (fromIndex, toIndex) =
  if index = fromIndex then toIndex
  elif index = toIndex then fromIndex
  else index

let private swap indexPair lst =
  Array.permute (fun index -> swapIndexMap index indexPair) lst

let shuffleInPlace fRand (lst: 'a []) =
  let swap pair =
    let temp = lst.[fst pair]
    lst.[fst pair] <- lst.[snd pair]
    lst.[snd pair] <- temp

  swapIndexPairs fRand lst |> Seq.iter swap
  lst
